src/main/java/wraith/fwaystones/mixin/MapStateMixin.java

```diff
@@ -0,0 +1,108 @@
+        MapState mapState = cir.getReturnValue();
+        NbtList nbtList = nbt.getList("waystones", NbtElement.COMPOUND_TYPE);
+        for (int k = 0; k < nbtList.size(); ++k) {
+            MapWaystoneMarker mapWaystoneMarker = MapWaystoneMarker.fromNbt(nbtList.getCompound(k));
+            ((MapStateMixin)(Object)mapState).fwaystones$waystones.put(mapWaystoneMarker.getKey(), mapWaystoneMarker);
+            ((MapStateMixin)(Object)mapState).addIcon(mapWaystoneMarker.getIconType(), null, mapWaystoneMarker.getKey(), mapWaystoneMarker.getPos().getX(), mapWaystoneMarker.getPos().getZ(), 180.0, mapWaystoneMarker.getName());
+        }
+    }
+
+    @Inject(method = "writeNbt", at = @At("TAIL"))
+    private void saveWaystonesNbt(NbtCompound nbt, CallbackInfoReturnable<NbtCompound> cir) {
+        NbtList nbtList = new NbtList();
+        for (MapWaystoneMarker mapwaystoneMarker : this.fwaystones$waystones.values()) {
+            nbtList.add(mapwaystoneMarker.getNbt());
+        }
+        nbt.put("waystones", nbtList);
+    }
+
+    @Inject(method = "copy", at = @At(value = "INVOKE", target = "Ljava/util/Map;putAll(Ljava/util/Map;)V", ordinal = 1))
+    private void copyWaystones(CallbackInfoReturnable<MapState> cir) {
+        ((MapStateMixin)(Object)cir.getReturnValue()).fwaystones$waystones.putAll(this.fwaystones$waystones);
+    }
+
+    @Override
+    public boolean addWaystone(WorldAccess world, BlockPos pos) {
+        double d = (double)pos.getX() + 0.5;
+        double e = (double)pos.getZ() + 0.5;
+        int i = 1 << ((MapState)(Object)this).scale;
+        double f = (d - (double)((MapState)(Object)this).centerX) / (double)i;
+        double g = (e - (double)((MapState)(Object)this).centerZ) / (double)i;
+        if (f >= -63.0 && g >= -63.0 && f <= 63.0 && g <= 63.0) {
+            MapWaystoneMarker mapWaystoneMarker = MapWaystoneMarker.fromWorldBlock(world, pos);
+            if (mapWaystoneMarker == null) {
+                return false;
+            }
+            if (this.fwaystones$waystones.remove(mapWaystoneMarker.getKey(), mapWaystoneMarker)) {
+                removeIcon(mapWaystoneMarker.getKey());
+                return true;
+            }
+            if (!((MapState)(Object)this).method_37343(256)) {
+                this.fwaystones$waystones.put(mapWaystoneMarker.getKey(), mapWaystoneMarker);
+                addIcon(mapWaystoneMarker.getIconType(), world, mapWaystoneMarker.getKey(), d, e, 180.0, mapWaystoneMarker.getName());
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @SuppressWarnings({"TooBroadScope", "UnclearExpression"})
+    @Override
+    public @Nullable BlockPos removeWaystone(BlockView world, int x, int z) {
+        Iterator<MapWaystoneMarker> iterator = this.fwaystones$waystones.values().iterator();
+        while (iterator.hasNext()) {
+            MapWaystoneMarker mapWaystoneMarker2;
```

Comment:
Is there a reason for this? It's not used anywhere (you have it in the equals check, but it does nothing).

Reply:
It's from the original minecraft code, I never bothered to remove it to not break anything.